src/psyclone/psyir/backend/sympy_writer.py

```diff
@@ -343,3 +508,147 @@ def is_intrinsic(self, operator):
+    def range_node(self, node):
+        '''This method is called when a Range instance is found in the PSyIR
+        tree. This implementation converts a range into three parameters
+        for the corresponding SymPy function.
+
+        :param node: a Range PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.Range`
+
+        :returns: the Fortran code as a string.
+        :rtype: str
+
+        '''
+        if node.parent and node.parent.is_lower_bound(
+                node.parent.indices.index(node)):
+            # The range starts for the first element in this
+            # dimension. This is the default in Fortran so no need to
+            # output anything.
+            start = self.lower_bound_name
+        else:
+            start = self._visit(node.start)
+
+        if node.parent and node.parent.is_upper_bound(
+                node.parent.indices.index(node)):
+            # The range ends with the last element in this
+            # dimension. This is the default in Fortran so no need to
+            # output anything.
+            stop = self.upper_bound_name
+        else:
+            stop = self._visit(node.stop)
+        result = f"{start},{stop}"
+
+        if isinstance(node.step, Literal) and \
+                node.step.datatype.intrinsic == \
+                ScalarType.Intrinsic.INTEGER and \
+                node.step.value == "1":
+            result += ",1"
+            # Step is 1. This is the default in Fortran so no need to
+            # output any text.
+        else:
+            step = self._visit(node.step)
+            result += f",{step}"
+
+        return result
+
+    # -------------------------------------------------------------------------
+    def sympy_to_psyir(self, sympy_expr, symbol_table):
+        '''This function converts a SymPy expression back into PSyIR. It first
+        parses the SymPy expression back into PSyIR, and then replaces all
+        array indices back into the corresponding Fortran values (since they
+        were replaced with three parameters to support array expressions), e.g.
+        ``a(i,i,1)`` will be converted back to ``a(i)``, and ``a(-inf,5,2)``
+        will become ``a(:5:2)``.
+
+        :param sympy_expr: the original SymPy expression.
+        :type sympy_expr: :py:class:`sympy.core.basic.Basic`
+        :param symbol_table: the symbol table required for parsing, it \
+            should be the table from which the original SymPy expression \
+            was created from (i.e. contain all the required symbols in the \
+            SymPy expression).
+        :type symbol_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+
+        :returns: the PSyIR representation of the SymPy expression.
+        :rtype: :py:class:`psyclone.psyir.nodes.Node`
+
+        '''
+        # Convert the new sympy expression to PSyIR
+        reader = FortranReader()
+        return reader.psyir_from_expression(str(sympy_expr), symbol_table)
```

Comment:
Sorry for being dense but I really can't see how this call to the standard Fortran reader does the conversions described in the docstring of this routine. What am I missing? Also, although it's only a small function, I think it should really be a separate PSyIR frontend.

Reply:
Done, and as explanation: SymPy does wonderful things in the background when converting things to a string. So a LOT is happening inside the `str(sympy_expr)`, including calls to the newly added Sympy-Array/Function writing method.

Reply:
Aha! Light is beginning to dawn...